### PR TITLE
[IMP] JWEditor: move renderer and contentEditable in a plugin

### PR DESCRIPTION
**Pull request by Gorash**

### CODE REVIEW

File: packages/core/src/JWEditor.ts [hunk 1]
```diff
@@ -15,72 +15,61 @@ export interface JWEditorConfig {
...
         // CorePlugin is a special mandatory plugin that handles the matching
         // between the core commands and the VDocument.
         this.addPlugin(CorePlugin);
 
+        // TODO: remove mandatory dependencie
+        this.addPlugin(RendererContentEditable);
+
+        if (this.target) {
+            // The original editable node is hidden until the editor stops.
+            this.target.style.display = 'none';
```
Comment (Goaman) [inline packages/core/src/JWEditor.ts]: detach the original node

File: packages/core/src/JWEditor.ts [hunk 2]
```diff
@@ -15,72 +15,61 @@ export interface JWEditorConfig {
...
+        if (this.target) {
+            if (this.target.tagName === 'TEXTAREA') {
+                const temp = document.createElement('jw-temp');
+                temp.innerHTML = (this.target as HTMLTextAreaElement).value;
+                this.vDocument = this.parser.parse(temp);
+            } else {
+                // Parse the editable in the internal format of the editor.
+                this.vDocument = this.parser.parse(this.target);
+            }
+        }
```
Comment (Goaman) [inline packages/core/src/JWEditor.ts]: move the logic in the caller of the start?

File: packages/plugin-devtools/test/devtools.test.ts [hunk 3]
```diff
@@ -746,7 +747,12 @@ describe('plugin-devtools', () => {
             await click(button);
 
             const subpanel = devtools.querySelector('devtools-panel.active mainpane-contents');
-            expect(subpanel.textContent).to.equal('insertTextsetSelection');
+            expect(
+                [].map.call(
+                    subpanel.querySelectorAll('td'),
+                    (td: HTMLTableCellElement) => td.textContent,
+                ),
```
Comment (Goaman) [inline packages/plugin-devtools/test/devtools.test.ts]:
```typescript
[...subpanel.querySelectorAll('td')].map((td: HTMLTableCellElement) => td.textContent)
```

File: packages/core/src/Parser.ts [hunk 4]
```diff
@@ -232,7 +239,11 @@ export class Parser {
      * @param container
      * @param offset
      */
-    _locate(container: Node, offset: number): [VNode, RelativePosition] {
+    _locate(
+        vDocumentMap: VDocumentMap,
+        container: Node,
+        offset: number,
+    ): [VNode, RelativePosition] {
```
Comment (dmo-odoo) [inline packages/core/src/Parser.ts]: Same for the Parser, I would gave the parser its own map rather than pollute the signature of all functions to pass the map to them.

File: packages/core/src/VDocumentMap.ts [hunk 5]
```diff
@@ -1,41 +1,42 @@
 import { Format } from '../../utils/src/Format';
 import { VNode } from './VNodes/VNode';
 
-const fromDom = new Map<Node, VNode[]>();
-const toDom = new Map<VNode, [Node, number]>();
-
-export const VDocumentMap = {
+export class VDocumentMap {
+    _fromDom = new Map<Node, VNode[]>();
+    _toDom = new Map<VNode, [Node, number]>();
```
Comment (dmo-odoo) [inline packages/core/src/VDocumentMap.ts]: I would have completely replaced the `VDocumentMap` by a map `Map<Node, VNode[]>` used only by the parser and a map `Map<VNode, [Node, number]>` use only by the renderer.

File: packages/core/src/JWEditor.ts [hunk 6]
```diff
@@ -15,40 +15,26 @@ export interface JWEditorConfig {
...
     /**
      * Start the editor on the editable DOM node set on this editor instance.
      */
-    async start(): Promise<void> {
-        // Parse the editable in the internal format of the editor.
-        this.vDocument = this.parser.parse(this._originalEditable);
-
-        // The original editable node is hidden until the editor stops.
-        this._originalEditable.style.display = 'none';
+    async start(value: Node | string = ''): Promise<void> {
```
Comment (dmo-odoo) [inline packages/core/src/JWEditor.ts]: I think this commit is completely out of scope of what we need for the renderer.

File: packages/core/src/Parser.ts [hunk 7]
```diff
@@ -132,6 +133,7 @@ export class Parser {
         // Start with the first child and the whole tree will be parsed.
         if (node.childNodes.length) {
             let currentContext: ParsingContext = {
+                parseNode: this.parseNode.bind(this, vDocumentMap),
```
Comment (dmo-odoo) [inline packages/core/src/Parser.ts]: @Zinston This is how Christophe managed to pass the "generic" parse function to the others. When I left yesterday you were going to try something along the line of `this.editor.parse('dom', vnode)`, did it work ? (Well, technically it was for the renderer but you got my point.)

File: packages/plugin-contentEditable/EventManager.ts [hunk 8]
```diff
@@ -1,19 +1,22 @@
...
 }
 
 export class EventManager {
     editor: JWEditor;
+    renderer: ContentEditable;
     eventNormalizer: EventNormalizer;
 
-    constructor(editor: JWEditor) {
-        this.editor = editor;
+    constructor(renderer: ContentEditable) {
```
[the remark this answers is not in the document]
Comment (Gorash) [reply packages/plugin-contentEditable/EventManager.ts]: no

File: packages/plugin-dom/DomRenderer.ts [hunk 9]
```diff
@@ -1,23 +1,30 @@
-import { Renderer } from '../core/src/Renderer';
 import { VDocumentMap } from '../core/src/VDocumentMap';
 import { VDocument } from '../core/src/VDocument';
 import { VNode, RelativePosition, Predicate } from '../core/src/VNodes/VNode';
```
Comment (Zinston) [inline packages/plugin-dom/DomRenderer.ts]: Remove unused imports.

File: packages/plugin-dom/DomRenderer.ts [hunk 10]
```diff
@@ -1,23 +1,30 @@
...
 import { VElement } from '../core/src/VNodes/VElement';
+import { RenderingFunctions } from '../core/src/RenderManager';
+import JWEditor from '../core/src/JWEditor';
+
+export type RenderingDOMFunction = (
+    currentContext: DomRenderingContext,
+) => [DomRenderingContext, DomRenderingMap];
 
 export interface DomRenderingContext {
+    vDocumentMap: VDocumentMap; // Mapping between vNodes and Nodes
```
[the remark this answers is not in the document]
Comment (Zinston) [reply packages/plugin-dom/DomRenderer.ts]: + `VDocumentMap refers to a value, but is being used as a type here.`

File: packages/plugin-dom/DomRenderer.ts [hunk 11]
```diff
@@ -1,23 +1,30 @@
...
 
-export class DomRenderer extends Renderer<
-    DomRenderingContext,
-    [DomRenderingContext, DomRenderingMap]
-> {
-    readonly id = 'dom';
-    _contextStack: Array<DomRenderingContext> = [];
+export class DomRenderer {
+    _contextStack: DomRenderingContext[] = [];
+    _renderingFunctions: RenderingFunctions;
```
[the remark this answers is not in the document]
Comment (Zinston) [reply packages/plugin-dom/DomRenderer.ts]: `Set<RenderingFunction>`

File: packages/plugin-list/test/List.test.ts [hunk 12]
```diff
@@ -281,9 +284,11 @@ describe('plugin-list', () => {
             ol.append(p9);
             motherList.append(ol);
 
-            editor.renderers.dom.render(editor.vDocument, editor.editable);
+            const renderer = new DomRenderer(editor);
+            const container = document.createElement('container');
```
[the remark this answers is not in the document]
Comment (Zinston) [reply packages/plugin-list/test/List.test.ts]: Fragment

File: packages/core/test/Parser.test.ts [hunk 13]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable max-nested-callbacks */
```
Comment (Zinston) [inline packages/core/test/Parser.test.ts]: Nope.

**Comment by dmo-odoo** [comment]
I'm closing this PR as it is replaced by this newer version: https://github.com/odoo-dev/jabberwock/pull/120.
